### PR TITLE
FAB-18192 Fixed TLS certs validation for consenters.

### PR DESCRIPTION
**Pull request by kopaygorodsky**
Signed-off-by: kopaygorodsky <vlad.kopaygorodsky@gmail.com>

#### Type of change

- Bug fix

#### Description
Fixes  usecase when you add new org with a consenter. 
Verification of consenter's TLS certs against simulated config, not the last one. To achieve that, metadata validator interface was changed, now it requires orderer config instead of just consensus metadata. Also, TLS verification was moved to VerifyConfigMetadata function. Added `ignoreCertExpiration` option to ignore expiration errors when validating config metadata. 

#### Additional details
1.`consenter.IsChannelMember` returns an error if metadata has a consenter with an invalid certificate or signed by unknown MSP. 
2.Added new tests,  fixed existing tests.
3.Checked the fix on a test network. 

#### Related issues

https://jira.hyperledger.org/browse/FAB-18192

### CODE REVIEW

File: integration/raft/config_test.go [hunk 1]
```diff
@@ -438,7 +443,7 @@ var _ = Describe("EndToEnd reconfiguration and onboarding", func() {
 			)
 			sess = nwo.UpdateOrdererConfigSession(network, orderer, network.SystemChannel.Name, current, updated, peer, orderer)
 			Eventually(sess, network.EventuallyTimeout).Should(gexec.Exit(1))
-			Expect(sess.Err).To(gbytes.Say(fmt.Sprintf("BAD_REQUEST -- error applying config update to existing channel 'systemchannel': consensus metadata update for channel config update is invalid: verifying tls client cert with serial number %d: x509: certificate signed by unknown authority", client.TLSCert.SerialNumber)))
+			Expect(sess.Err).To(MatchError(ContainSubstring(fmt.Sprintf("BAD_REQUEST -- error applying config update to existing channel 'systemchannel': consensus metadata update for channel config update is invalid: invalid new config metadadata: verifying tls client cert with serial number %d: x509: certificate signed by unknown authority", certX509.SerialNumber))))
```
Comment (wlahti) [inline integration/raft/config_test.go]:
```suggestion
			Expect(sess.Err).To(MatchError(ContainSubstring(fmt.Sprintf("BAD_REQUEST -- error applying config update to existing channel 'systemchannel': consensus metadata update for channel config update is invalid: invalid new config metadata: verifying tls client cert with serial number %d: x509: certificate signed by unknown authority", certX509.SerialNumber))))
```

File: internal/cryptogen/msp/msp.go [hunk 2]
```diff
@@ -249,7 +249,7 @@ func GenerateVerifyingMSP(
 	return nil
 }
 
-func createFolderStructure(rootDir string, local bool) error {
+func CreateFolderStructure(rootDir string, local bool) error {
```
Comment (wlahti) [inline internal/cryptogen/msp/msp.go]: Not sure we should make this public just to use it for tests. Looking at the details of the function, why not just copy this as a helper function to the test file(s) you need to create your own msp?

File: integration/raft/config_test.go [hunk 3]
```diff
@@ -207,7 +209,7 @@ var _ = Describe("EndToEnd reconfiguration and onboarding", func() {
 			exitCode := network.CreateChannelExitCode(channel, orderer, org1Peer0, org1Peer0, org2Peer0, orderer)
 			Expect(exitCode).NotTo(Equal(0))
 			Consistently(process.Wait).ShouldNot(Receive()) // malformed tx should not crash orderer
-			Expect(runner.Err()).To(gbytes.Say(`invalid new config metdadata: ElectionTick \(10\) must be greater than HeartbeatTick \(10\)`))
+			Expect(runner.Err()).To(gbytes.Say(`invalid new config metadadata: ElectionTick \(10\) must be greater than HeartbeatTick \(10\)`))
```
Comment (wlahti) [inline integration/raft/config_test.go]:
```suggestion
			Expect(runner.Err()).To(gbytes.Say(`invalid new config metadata: ElectionTick \(10\) must be greater than HeartbeatTick \(10\)`))
```

File: integration/raft/config_test.go [hunk 4]
```diff
@@ -235,7 +237,7 @@ var _ = Describe("EndToEnd reconfiguration and onboarding", func() {
 
 			sess := nwo.UpdateOrdererConfigSession(network, orderer, channel, config, updatedConfig, org1Peer0, orderer)
 			Eventually(sess, network.EventuallyTimeout).Should(gexec.Exit(1))
-			Expect(sess.Err).To(gbytes.Say(`invalid new config metdadata: ElectionTick \(10\) must be greater than HeartbeatTick \(10\)`))
+			Expect(sess.Err).To(gbytes.Say(`invalid new config metadadata: ElectionTick \(10\) must be greater than HeartbeatTick \(10\)`))
```
Comment (wlahti) [inline integration/raft/config_test.go]:
```suggestion
			Expect(sess.Err).To(gbytes.Say(`invalid new config metadata: ElectionTick \(10\) must be greater than HeartbeatTick \(10\)`))
```

File: orderer/consensus/etcdraft/chain.go [hunk 5]
```diff
@@ -1288,27 +1289,33 @@ func (c *Chain) newConfigMetadata(block *common.Block) *etcdraft.ConfigMetadata
...
 
-	err := CheckConfigMetadata(newMetadata)
+	verifyOpts, err := createX509VerifyOptions(oldOrdererConfig, newOrdererConfig)
 	if err != nil {
-		return errors.Wrap(err, "invalid new config metdadata")
+		return errors.Wrapf(err, "failed to create x509 verify options from old and new orderer config")
+	}
+
+	if err := VerifyConfigMetadata(newMetadata, verifyOpts); err != nil {
+		return errors.Wrap(err, "invalid new config metadadata")
```
Comment (wlahti) [inline orderer/consensus/etcdraft/chain.go]:
```suggestion
		return errors.Wrap(err, "invalid new config metadata")
```

File: orderer/consensus/etcdraft/consenter_test.go [hunk 6]
```diff
@@ -151,10 +157,35 @@ var _ = Describe("Consenter", func() {
...
+			//Adding tls ca root cert to an existing ordering org's MSP definition. Don't want to spoil sampleconfig, copying it to some tmp dir.
+			mspDir, err := ioutil.TempDir("", "msp-")
+			Expect(err).NotTo(HaveOccurred())
+			Expect(confAppRaft.Orderer).NotTo(BeNil())
+			Expect(confAppRaft.Orderer.Organizations).To(HaveLen(1))
+
+			err = testutil.CopyDir(confAppRaft.Orderer.Organizations[0].MSPDir, mspDir, true)
+			Expect(err).NotTo(HaveOccurred())
+			//confAppRaft.Orderer.Organizations[0].MSPDir = mspDir
+			//confAppRaft.Orderer.Organizations[0].ID = fmt.Sprintf("SampleMSP-%d", time.Now().UnixNano())
```
Comment (wlahti) [inline orderer/consensus/etcdraft/consenter_test.go]: Nit: remove?

File: orderer/consensus/etcdraft/util.go [hunk 7]
```diff
@@ -198,7 +201,7 @@ func ConsensusMetadataFromConfigBlock(block *common.Block) (*etcdraft.ConfigMeta
 }
 
 // CheckConfigMetadata validates Raft config metadata
```
Comment (wlahti) [inline orderer/consensus/etcdraft/util.go]:
```suggestion
// VerifyConfigMetadata validates Raft config metadata
```
Nit

File: orderer/consensus/etcdraft/util.go [hunk 8]
```diff
@@ -253,16 +253,165 @@ func CheckConfigMetadata(metadata *etcdraft.ConfigMetadata) error {
...
+		Roots:         tlsRoots,
+		Intermediates: tlsIntermediates,
+		KeyUsages: []x509.ExtKeyUsage{
+			x509.ExtKeyUsageClientAuth,
+			x509.ExtKeyUsageServerAuth,
+		},
+	}, nil
+}
+
+func ValidateConsenterTLSCerts(c *etcdraft.Consenter, opts *x509.VerifyOptions) error {
```
Comment (wlahti) [inline orderer/consensus/etcdraft/util.go]: Any need for this to be public?

File: orderer/consensus/etcdraft/util.go [hunk 9]
```diff
@@ -273,6 +422,8 @@ type ConsenterCertificate struct {
 	Logger               *flogging.FabricLogger
 }
 
+// type ConsenterCertificate []byte
```
Comment (wlahti) [inline orderer/consensus/etcdraft/util.go]: Nit: remove?

File: orderer/consensus/etcdraft/validator_test.go [hunk 10]
```diff
@@ -82,39 +96,46 @@ var _ = Describe("Metadata Validation", func() {
...
-				chain.ValidateConsensusMetadata([]byte("test"), []byte("test"), false)
-			}).To(Panic())
-		})
-
-		It("fails when new consensus metadata is not well-formed", func() {
-			oldBytes, _ := proto.Marshal(&etcdraftproto.ConfigMetadata{})
-			Expect(chain.ValidateConsensusMetadata(oldBytes, []byte("test"), false)).NotTo(Succeed())
-		})
-	})
+	//When("determining parameter well-formedness", func() {
```
Comment (wlahti) [inline orderer/consensus/etcdraft/validator_test.go]: What's going on here? :) What happened to these tests?

**Review by wlahti — COMMENTED** [review]
Thanks for picking this task up! A few general comments that stood out on a first pass. 

Going to take a deeper look at the more substantial changes in etcdraft/util.go.

File: internal/cryptogen/msp/msp.go (hunk 2 above)
Comment (kopaygorodsky) [reply internal/cryptogen/msp/msp.go]: yep, not a problem, will make it private again.

File: orderer/consensus/etcdraft/consenter_test.go (hunk 6 above)
Comment (kopaygorodsky) [reply orderer/consensus/etcdraft/consenter_test.go]: I'm in a process of debugging it 
https://chat.hyperledger.org/channel/fabric-orderer?msg=vjDwFDFtGvhWHKzKB here is the problem. 
will cleanup & remove ofc

File: orderer/consensus/etcdraft/validator_test.go (hunk 10 above)
Comment (kopaygorodsky) [reply orderer/consensus/etcdraft/validator_test.go]: don't worry, I'm in the middle of fixing them. They pass on my local machine, but not on CI.  Everything will be cleaned and fixed.

File: orderer/consensus/etcdraft/util.go [hunk 11]
```diff
@@ -197,8 +200,10 @@ func ConsensusMetadataFromConfigBlock(block *common.Block) (*etcdraft.ConfigMeta
 	return MetadataFromConfigUpdate(configUpdate)
 }
 
-// CheckConfigMetadata validates Raft config metadata
-func CheckConfigMetadata(metadata *etcdraft.ConfigMetadata) error {
+// VerifyConfigMetadata validates Raft config metadata.
+// If x509.VerifyOpts is nil, it will do only sanity check of certificates.
+// If ignoreCertExpiration is true, it will verify certificate and ignore expiration errors.
+func VerifyConfigMetadata(metadata *etcdraft.ConfigMetadata, verifyOpts *x509.VerifyOptions, ignoreCertExpiration bool) error {
```
Comment (yacovm) [inline orderer/consensus/etcdraft/util.go]: Why would we want to ignore expiration?

File: orderer/consensus/etcdraft/util.go [hunk 12]
```diff
@@ -253,16 +255,176 @@ func CheckConfigMetadata(metadata *etcdraft.ConfigMetadata) error {
...
+			return certificateList, err
+		}
+
+		certificateList = append(certificateList, certificate)
+	}
+
+	return certificateList, nil
+}
+
+func getCertFingerprint(cert *x509.Certificate) (string, error) {
```
Comment (yacovm) [inline orderer/consensus/etcdraft/util.go]: Why do you need to "compress" the certificate into a short string? You use it as a key in a map that is actually used as a set. 

You can just "cast" the certificate to string instead and check if the certificate (as a string) itself exits in a map. No need to hash it. 

```
set := make(map[string]struct{})
_, exists := set[string(certificate)]
```

File: orderer/consensus/etcdraft/util.go [hunk 13]
```diff
@@ -253,16 +255,176 @@ func CheckConfigMetadata(metadata *etcdraft.ConfigMetadata) error {
...
+				return nil, errors.Wrap(err, "parsing tls intermediate certs")
+			}
+
+			for _, cert := range rootCerts {
+				fingerprint, err := getCertFingerprint(cert)
+				if err != nil {
+					return nil, errors.WithStack(err)
+				}
+
+				if _, exists := tlsRootsMap[fingerprint]; !exists {
```
Comment (yacovm) [inline orderer/consensus/etcdraft/util.go]: What problem are you trying to avoid, by not adding the certificate twice?

File: orderer/consensus/etcdraft/util.go [hunk 14]
```diff
@@ -253,16 +255,176 @@ func CheckConfigMetadata(metadata *etcdraft.ConfigMetadata) error {
...
+}
+
+func createX509VerifyOptions(oldOrdererConfig, newOrdererConfig channelconfig.Orderer) (*x509.VerifyOptions, error) {
+	tlsRoots := x509.NewCertPool()
+	tlsIntermediates := x509.NewCertPool()
+
+	tlsRootsMap := make(map[string]*x509.Certificate)
+	tlsIntermediatesMap := make(map[string]*x509.Certificate)
+
+	if oldOrdererConfig != nil {
```
Comment (yacovm) [inline orderer/consensus/etcdraft/util.go]: I am confused. Why do we care about the old config at all? I thought we only care about the new config.

File: orderer/consensus/etcdraft/util.go [hunk 15]
```diff
@@ -253,16 +255,176 @@ func CheckConfigMetadata(metadata *etcdraft.ConfigMetadata) error {
...
+	if err != nil {
+		return errors.Wrapf(err, "parsing tls client cert of %s:%d", c.Host, c.Port)
+	}
+
+	serverCert, err := parseCertificateFromBytes(c.ServerTlsCert)
+	if err != nil {
+		return errors.Wrapf(err, "parsing tls server cert of %s:%d", c.Host, c.Port)
+	}
+
+	if opts != nil {
```
Comment (yacovm) [inline orderer/consensus/etcdraft/util.go]: If we have nil options, then aren't we badly screwed?

File: orderer/consensus/etcdraft/util.go [hunk 16]
```diff
@@ -253,16 +255,176 @@ func CheckConfigMetadata(metadata *etcdraft.ConfigMetadata) error {
...
+		verifyOpts := *opts
+
+		var verify = func(certType string, cert *x509.Certificate, opts x509.VerifyOptions) error {
```
Comment (yacovm) [inline orderer/consensus/etcdraft/util.go]: If any certificate across the validation chain is expired, then why is this OK? What is the use case you are trying to take into account?

File: orderer/consensus/etcdraft/util.go [hunk 17]
```diff
@@ -253,16 +255,176 @@ func CheckConfigMetadata(metadata *etcdraft.ConfigMetadata) error {
...
+			if _, err := clientCert.Verify(verifyOpts); err != nil {
+				if validationRes, ok := err.(x509.CertificateInvalidError); !ok || (!ignoreExpiration || validationRes.Reason != x509.Expired) {
```
Comment (yacovm) [inline orderer/consensus/etcdraft/util.go]: There is a better way of ignoring expiration, look [here](https://github.com/hyperledger/fabric/blob/master/msp/mspimplvalidate.go#L284).

File: orderer/consensus/etcdraft/membership.go [hunk 18]
```diff
@@ -112,94 +105,6 @@ func ComputeMembershipChanges(oldMetadata *etcdraft.BlockMetadata, oldConsenters
...
-
-	_, err = serverCert.Verify(opts)
-	if err != nil {
-		return fmt.Errorf("verifying tls server cert with serial number %d: %v", serverCert.SerialNumber, err)
-	}
-
-	return nil
-}
-
-func createX509VerifyOptions(orgs map[string]channelconfig.OrdererOrg) (x509.VerifyOptions, error) {
```
Comment (yacovm) [inline orderer/consensus/etcdraft/membership.go:142]: why did you move these to a different file?

File: orderer/consensus/etcdraft/util.go (hunk 11 above)
Comment (kopaygorodsky) [reply orderer/consensus/etcdraft/util.go]: Some consenters, for which my org is not responsible for, could have expired certificates. The responsible org rotates certificates, but why should I wait for it?  It should be possible to do a config update if we have enough nodes for a quorum.

File: orderer/consensus/etcdraft/util.go (hunk 14 above)
Comment (kopaygorodsky) [reply orderer/consensus/etcdraft/util.go:312]: hm, true.

File: orderer/consensus/etcdraft/membership.go (hunk 18 above)
Comment (kopaygorodsky) [reply orderer/consensus/etcdraft/membership.go:142]: `ComputeMemebershipChanges` shouldn't be responsible for TLS validations, that's why I moved it.  Those functions should be near VerifyConfigMetadata.

File: orderer/consensus/etcdraft/util.go (hunk 13 above)
Comment (kopaygorodsky) [reply orderer/consensus/etcdraft/util.go]: I just wanted this method to be efficient, I'll revert this part, np.

File: orderer/consensus/etcdraft/util.go (hunk 12 above)
Comment (kopaygorodsky) [reply orderer/consensus/etcdraft/util.go]: will fix

File: orderer/consensus/etcdraft/util.go (hunk 11 above)
Comment (kopaygorodsky) [reply orderer/consensus/etcdraft/util.go]: This [usecase](https://github.com/hyperledger/fabric/blob/master/integration/raft/cft_test.go#L519) fails if I take into account the expiration of certificates.

File: orderer/consensus/etcdraft/util.go (hunk 15 above)
Comment (kopaygorodsky) [reply orderer/consensus/etcdraft/util.go]: `validateConsenterTLSCerts` could be used for sanity check(parsing x509) if `opts` is nil.

File: orderer/consensus/etcdraft/util.go (hunk 16 above)
Comment (kopaygorodsky) [reply orderer/consensus/etcdraft/util.go]: I tried to answer this in the first comment above. There are certificates for which I'm not responsible when a consenter is managed by another org. They can be expired for some time. If we have enough nodes for a quorum - we should be able to perform an update.

File: orderer/consensus/etcdraft/util.go (hunk 17 above)
Comment (kopaygorodsky) [reply orderer/consensus/etcdraft/util.go]: that's what I did in a first-place actually, but there is [usecase](https://github.com/hyperledger/fabric/blob/master/integration/raft/cft_test.go#L519)  that fails if I set `opts.CurrentTime` to client's or server's `NotBefore`. In that use case CA cert is also expired(somewhere in the test before), so I should compare the whole chain, find the oldest one, and set `opts.CurrentTime = oldest.NotBefore`. The amount of code/logic is much bigger and I thought the way of ignoring expiration is not relevant.

File: orderer/consensus/etcdraft/util.go (hunk 15 above)
Comment (kopaygorodsky) [reply orderer/consensus/etcdraft/util.go]: fixed.

File: orderer/consensus/etcdraft/util.go (hunk 14 above)
Comment (kopaygorodsky) [reply orderer/consensus/etcdraft/util.go]: @yacovm this [use case](https://github.com/hyperledger/fabric/blob/master/integration/raft/config_test.go#L393) fails if I do not add certs from the old config. TLS CA cert is removed from MSP, but it's consenters are still in the config. 
Error: `certificate signed by unknown authority...`
I deleted the use case because it's not relevant anymore.

File: orderer/consensus/etcdraft/util.go (hunk 11 above)
Comment (yacovm) [reply orderer/consensus/etcdraft/util.go]: In that case, I think it makes sense to only look at the delta (new consenters added) and not at all certificates.

File: orderer/consensus/etcdraft/util.go (hunk 16 above)
Comment (yacovm) [reply orderer/consensus/etcdraft/util.go]: Then we should only care about the delta

File: orderer/consensus/etcdraft/util.go (hunk 11 above)
Comment (kopaygorodsky) [reply orderer/consensus/etcdraft/util.go]: `VerifyConfigMetadata` is used in `consenter.IsChannelMemeber` [here](https://github.com/hyperledger/fabric/pull/1888/files#diff-31c50bd6bdb92751de2354b5ac74099eR274), there is no delta, it detects whether it implies that this orderer is a member of the channel or no, expiration should be ignored.
Delta is validated only during config update, [here](https://github.com/hyperledger/fabric/pull/1888/files#diff-9f34970b484b953e4f5ff2b4b98397e9R1344).
Maybe, I should ignore expiration by default, remove `ignoreCertExpiration` option, don't see any place where it could be set to false. And validate delta set using `validateConsenterTLSCerts` after membership is calculated.

Comment (kopaygorodsky) [reply orderer/consensus/etcdraft/util.go]: maybe I can add `MembershipChanges` as optional param to `VerifyConfigMetadata`, in that way I can verify expiration only on new consenters if there are any.

File: orderer/consensus/etcdraft/membership.go (hunk 18 above)
Comment (jyellick) [reply orderer/consensus/etcdraft/membership.go:142]: I agree that `ComputeMembershipChanges` should not be responsible for TLS validations.  However, for future reference, when performing operations like moving code which produces a large diff, it's much better to perform the move in its own PR, or minimally, in its own commit.  It makes reviewing changes much easier.

**Review by jyellick — COMMENTED** [review]
Thanks for this pull request! It's definitely a bug that we're not validating using the latest config.  There are a few points, in particular since we are looking at the full config, and not performing the validation only on the update information, I'm a little concerned about unrelated config updates being rejected because of temporal checks.  I've included a comment at the spot, but, I think we could simplify this code by simply skipping the expiration checks.  If we are really convinced that they are the right thing to do, then we need to be careful that we only apply them to new certificates in the channel config, and not existing ones.  If we apply it to existing ones, then as soon any orderer TLS cert expires, all channel updates would begin failing until the issue is fixed.

File: orderer/consensus/etcdraft/chain.go [hunk 19]
```diff
@@ -1288,34 +1289,40 @@ func (c *Chain) newConfigMetadata(block *common.Block) *etcdraft.ConfigMetadata
 
 // ValidateConsensusMetadata determines the validity of a
 // ConsensusMetadata update during config updates on the channel.
-func (c *Chain) ValidateConsensusMetadata(oldMetadataBytes, newMetadataBytes []byte, newChannel bool) error {
+func (c *Chain) ValidateConsensusMetadata(oldOrdererConfig, newOrdererConfig channelconfig.Orderer, newChannel bool) error {
 	// metadata was not updated
-	if newMetadataBytes == nil {
+	if newOrdererConfig == nil {
```
[the remark this answers is not in the document]
Comment (kopaygorodsky) [reply orderer/consensus/etcdraft/chain.go:1293]: fixed

File: orderer/consensus/etcdraft/util_test.go [hunk 20]
```diff
@@ -7,21 +7,27 @@ SPDX-License-Identifier: Apache-2.0
...
 	"github.com/hyperledger/fabric/protoutil"
+	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
+	"io/ioutil"
+	"path/filepath"
+	"testing"
+)
+
+const (
+	consentersTestDataDir = "testdata/consenters_certs/"
```
[the remark this answers is not in the document]
Comment (kopaygorodsky) [reply orderer/consensus/etcdraft/util_test.go:28]: fixed

File: orderer/consensus/etcdraft/chain.go [hunk 21]
```diff
@@ -1328,11 +1335,18 @@ func (c *Chain) ValidateConsensusMetadata(oldMetadataBytes, newMetadataBytes []b
...
 	dummyOldConsentersMap := CreateConsentersMap(dummyOldBlockMetadata, oldMetadata)
-	changes, err := ComputeMembershipChanges(dummyOldBlockMetadata, dummyOldConsentersMap, newMetadata.Consenters, c.support.SharedConfig())
+	changes, err := ComputeMembershipChanges(dummyOldBlockMetadata, dummyOldConsentersMap, newMetadata.Consenters)
 	if err != nil {
 		return err
 	}
 
+	//new config metadata was verified above. Additionally need to check new consenters for certificates expiration
+	for _, c := range changes.AddedNodes {
+		if err := validateConsenterTLSCerts(c, verifyOpts, false); err != nil {
```
[the remark this answers is not in the document]
Comment (kopaygorodsky) [reply orderer/consensus/etcdraft/chain.go:1356]: @jyellick but slice`changes.AddedNodes` contains only new nodes(only one node actually), doesn't it?

File: orderer/consensus/etcdraft/membership.go (hunk 18 above)
Comment (kopaygorodsky) [reply orderer/consensus/etcdraft/membership.go:142]: sorry for that, will take into account next time :)

File: orderer/consensus/etcdraft/chain.go (hunk 21 above)
[the remark this answers is not in the document]
Comment (jyellick) [reply orderer/consensus/etcdraft/chain.go:1356]: You're correct, sorry for missing this.

**Review by jyellick — APPROVED** [review]
I think this is in good enough shape to merge.  I've opened https://jira.hyperledger.org/browse/FAB-18257 to address removing the test fixtures once go v1.15 is being used for Fabric.

**Comment by wlahti** [comment]
@jyellick Can you run the backport command to get this in the release-2.2 branch? Thanks! (I assume only maintainers are allowed to do that based on my previous failed attempts to run it myself) 

@Mergifyio backport release-2.2

**Comment by jyellick** [comment]
@Mergifyio backport release-2.2

**Comment by kopaygorodsky** [comment]
woohoo, my first PR to fabric is merged :)

**Comment by robmurgai** [comment]
Congratulations, Nice work :)
                                                                                                                                                        S Rob Murgai
                                                He / Him / His
                        Program Director - Hyperledger Fabric 
                        IBM Blockchain
                        Raleigh-Durham, NC
                        M: 919.342.8432 | LinkedIn: Rob-Murgai                                                                                                                                                                                                
 
 
----- Original message -----From: Vladyslav Kopaihorodskyi <notifications@github.com>To: hyperledger/fabric <fabric@noreply.github.com>Cc: Subscribed <subscribed@noreply.github.com>Subject: [EXTERNAL] Re: [hyperledger/fabric] FAB-18192 Fixed TLS certs validation for consenters. (#1888)Date: Tue, Oct 6, 2020 1:42 PM 
 
woohoo, my first PR to fabric is merged :)

**Comment by lucianoacsilva** [comment]
Hello! Is that the new version with this latest fix already available in Docker Hub?

**Comment by lindluni** [comment]
There is not yet, you will need to build it yourself, or pull it from `hyperledger-fabric.jfrog.io/fabric-peer:amd64-2.2-stable` which is the unreleased artifact, but the current tip of the 2.2 branch. If you want the latest master you can use the `amd64-latest` tag instead.

**Comment by lucianoacsilva** [comment]
Hello, @btl5037 ,

I've pulled here the `hyperledger-fabric.jfrog.io/fabric-peer:amd64-2.2-stable`, as you mentioned, but the bug still persists when I try to update the config block with a new adding org:

`
Error: got unexpected status: BAD_REQUEST -- error applying config update to existing channel 'system-channel': consensus metadata update for channel config update is invalid: verifying tls client cert with serial number 84262169678602556738840170285785102293826956193: x509: certificate signed by unknown authority
`

Is this bug still persisting with you?

**Comment by lindluni** [comment]
@lucianoacsilva Can you pull the image again, artifactory was receiving, but not publishing the updates, it's published now.

**Comment by lucianoacsilva** [comment]
@btl5037 , removed the image, pulled it again when starting main network, error still persists:

`
Error: got unexpected status: BAD_REQUEST -- error applying config update to existing channel 'meat-trace': consensus metadata update for channel config update is invalid: verifying tls client cert with serial number 339235991777995742225504705667499867821380891148: x509: certificate signed by unknown authority
`

**Comment by lindluni** [comment]
@lucianoacsilva Can you ping me on chat.hyperledger.org, my ID is `BrettLogan`

**Comment by kopaygorodsky** [comment]
@lucianoacsilva @btl5037 I've built an image from master branch and this fix works. On the other hand, I've tried hyperledger-fabric.jfrog.io/fabric-orderer:amd64-2.2-stable and see this issue.

**Comment by lucianoacsilva** [comment]
I see, @kopaygorodsky . How could I build those images from the master branch?

**Comment by kopaygorodsky** [comment]
@lucassaldanha 
try `hyperledger-fabric.jfrog.io/fabric-orderer:amd64-latest`
or 
git clone fabric
make orderer-docker
and push docker image to your docker registry

**Comment by lucianoacsilva** [comment]
Wait, is that the image of the peer or the orderer? @btl5037 mentioned me the `hyperledger-fabric.jfrog.io/fabric-peer:amd64-2.2-stable` image, not the `hyperledger-fabric.jfrog.io/fabric-orderer:amd64-latest`one.

**Comment by kopaygorodsky** [comment]
orderer validates block, so you need to use orderer image
